\documentclass[article, shortnames]{jss}

%% -- LaTeX packages and custom commands --------------

%% recommended packages
\usepackage{thumbpdf,lmodern}

%% another package (only for this demo article)
\usepackage{framed}
\usepackage{subcaption} %continuefloat
\usepackage{xfrac}
\usepackage{mathtools}
\usepackage{natbib}
 
\usepackage[acronym, xindy, toc, acronym, style=super,  nogroupskip, nonumberlist, nopostdot]{glossaries}
\setlength{\glsdescwidth}{0.75\textwidth}
\include{glossary}
\makeglossaries

\usepackage{placeins}

%\usepackage[colorlinks,
%           linkcolor={black},
%           citecolor={black},
%           urlcolor={blue!80!white}]{hyperref}
           %allbordercolors={0 0 0},
           %pdfborderstyle={/S/U/W 1}
 
%% new custom commands
\newcommand{\class}[1]{`\code{#1}'}
\newcommand{\fct}[1]{\code{#1()}}


%% - Article metainformation (author, title, ...) ----------

%% - \author{} with primary affiliation
%% - \Plainauthor{} without affiliations
%% - Separate authors by \And or \AND (in \author) or by comma (in \Plainauthor).
%% - \AND starts a new line, \And does not.
\author{S. Thomas Kelly\\University of Otago
   \And Michael A. Black\\University of Otago}
\Plainauthor{Thomas Kelly, Michael Black}

%% - \title{} in title case
%% - \Plaintitle{} without LaTeX markup (if any)
%% - \Shorttitle{} with LaTeX markup (if any), used as running title
\title{\pkg{graphsim}: An \proglang{R} package for simulating gene expression data from graph structures of biological pathways}
\Plaintitle{graphsim: An R package for simulating gene expression data from graph structures of biological pathways}
\Shorttitle{Simulating gene expression from pathways in R}

%% - \Abstract{} almost as usual
\Abstract{
Transcriptomic analysis is used to capture the molecular state of a cell or sample in many biological and medical applications. In addition to identifying alterations in activity at the level of individual genes, understanding changes in the gene networks that regulate fundamental biological mechanisms is also an important objective of molecular analysis. As a result, databases that describe biological pathways are increasingly relied on to assist with the interpretation of results from large-scale genomics studies. Incorporating information from biological pathways and gene regulatory networks into a genomic data analysis is a popular strategy, and there are many methods that provide this functionality for gene expression data. When developing or comparing such methods, it is important to gain an accurate assessment of their performance, with simulation-based validation studies a popular choice.  This necessitates the use of simulated data that correctly accounts for pathway relationships and correlations. Here we present a versatile statistical framework to simulate correlated gene expression data from biological pathways, by sampling from a multivariate normal distribution derived from a graph structure. This procedure has been released as the \pkg{graphsim} \proglang{R} package (\url{https://github.com/TomKellyGenetics/graphsim}) and is compatible with any graph structure that can be described using the \pkg{igraph} package.
}

%% - \Keywords{} with LaTeX markup, at least one required
%% - \Plainkeywords{} without LaTeX markup (if necessary)
%% - Should be comma-separated and in sentence case.
\Keywords{gene-expression, simulation, genomics, pathway, network, \proglang{R}}
\Plainkeywords{gene-expression, simulation, genomics, pathway, network, R}

%% - \Address{} of at least one author
%% - May contain multiple affiliations for each author
%%   (in extra lines, separated by \emph{and}\\).
%% - May contain multiple authors for the same affiliation
%%   (in the same first line, separated by comma).
\Address{
  S. Thomas Kelly\\
  Division of Genomic Medicine
  RIKEN Center for Integrative Medical Sciences\\
  Yokohama, Japan \\
  E-mail: \email{tom.kelly@riken.jp}\\
  \emph{and}\\
  Department of Biochemistry\\
  University of Otago\\
  Dunedin, New Zealand\\
  %E-mail: \email{tom.kelly@postgrad.otago.ac.nz}%\\
  %URL: \url{https://eeecon.uibk.ac.at/~zeileis/}

  Michael Black\\
  Department of Biochemistry\\
  University of Otago\\
  PO Box 56\\
  Dunedin, New Zealand\\
  E-mail: \email{mik.black@otago.ac.nz}%\\
  %URL: \url{https://eeecon.uibk.ac.at/~zeileis/}
}

\begin{document}


%% - Introduction ---------------------

%% - In principle "as usual".
%% - But should typically have some discussion of both _software_ and _methods_.
%% - Use \proglang{}, \pkg{}, and \code{} markup throughout the manuscript.
%% - If such markup is in (sub)section titles, a plain text version has to be
%%   added as well.
%% - All software mentioned should be properly \cite-d.
%% - All abbreviations should be introduced.
%% - Unless the expansions of abbreviations are proper names (like "Journal
%%   of Statistical Software" above) they should be in sentence case (like
%%   "generalized linear models" below).

\section[Introduction: inference and modelling of biological networks]{Introduction: inference and modelling of biological networks} \label{sec:intro}

\shortcites{Perou2000, Arner2015, Reactome, Komatsu2013, Yamaguchi2007, Shimamura2009, Hirose2008, Li2015, Genz2016}

Network analysis of molecular biological pathways has the potential to lead to new insights into biology and medical genetics \citep{Barabasi2004, Hu2016}. Since gene expression profiles capture a consistent signature of the regulatory state of a cell \citep{Perou2000, Ozsolak2011, Svensson2018}, they can be used to analyse complex molecular states with genome-scale data. However, biological pathways are often analysed in a reductionist paradigm as amorphous sets of genes involved in particular functions, despite the fact that the relationships defined by pathway structure could further inform gene expression analyses. In many cases, the pathway relationships are well-defined, experimentally-validated, and are available in public databases \citep{Reactome}. 
%While pathway analysis over-representation analysis using the hypergeometric test is an effective approach to identify enriched functional groups \citep{Mooney2015, Werner2008}, it is limited by the accuracy and coverage of the biological databases used. Furthermore, pathway enrichment does not utilise the pathway structure information that has been gathered for many pathways and is supported by databases such as Reactome \citep{Reactome}. 
As a result, network analysis techniques could play an important role in furthering our understanding of biological pathways and aiding in the interpretation of genomics studies. %Furthermore with modelling and simulation procedures to evaluate them are needed.

Gene networks provide insights into how cells are regulated, by mapping regulatory interactions between target genes and transcription factors, enhancers, and sites of epigenetic marks or chromatin structures \citep{Barabasi2004, Yamaguchi2007}. Inference of these regulatory interactions for genomics investigations has the potential to radically expand the range of candidate biological pathways to be further explored, or to improve the accuracy of bioinformatics and functional genomic analysis. %which relies on these pathways being curated.
A number of methods have already been developed to utilise timecourse gene expression data \citep{Arner2015, Yamaguchi2007} using gene regulatory modules in state-space models and recursive vector autoregressive models \citep{Hirose2008, Shimamura2009}. Various approaches to gene regulation and networks at the genome-wide scale have lead to novel biological insights \citep{Arner2015, Komatsu2013}. However, inference of regulatory networks has thus far relied on experimental validation or resampling-based approaches to estimate the likelihood of specific network modules being predicted \citep{Markowetz2007,Hawe2019}.

There is a need, therefore, for a systematic framework for statistical modelling and simulation of gene expression data derived from hypothetical, inferred or known gene networks. Here we present an \proglang{R} package to achieve this, where samples from a multivariate normal distribution are used to generate normally-distributed log-expression data, with correlations between genes derived from the structure of the underlying pathway or gene regulatory network. This methodology enables simulation of expression profiles that approximate the log-transformed and normalised data from microarray and bulk or single-cell RNA-Seq experiments. This procedure has been released as the \pkg{graphsim} \proglang{R} package to enable the generation of simulated gene expression datasets containing pathway relationships from a known underlying network. These simulated datasets can be used to evaluate various bioinformatics methodologies, including statistical and network inference procedures.

%This procedure enables modelling of genes within a biological pathway and the investigation of the impact of pathway structure on identification of biomarkers from gene expression analysis techniques. It was developed for biological pathway members with correlated gene expression (higher than the background of genes in other pathways) but it may also be applicable to modelling protein levels (e.g, in a kinase regulation cascade) or substrates and products (e.g., in a metabolic pathway). The simulation procedure was designed to use graph structures (such as those shown in Figure~\ref{fig:simple_graph}) to inform development of a ``Sigma'' variance-covariance matrix ($\Sigma$) for sampling from a multivariate normal distribution. % (using the \pkg{mvtnorm} \proglang{R} package). 

\iffalse
\begin{leftbar}
The introduction is in principle ``as usual''. However, it should usually embed
both the implemented \emph{methods} and the \emph{software} into the respective
relevant literature. For the latter both competing and complementary software
should be discussed (within the same software environment and beyond), bringing
out relative (dis)advantages. All software mentioned should be properly
\verb|\cite{}|d. (See also Appendix~\ref{app:bibtex} for more details on
\textsc{Bib}{\TeX}.)

For writing about software JSS requires authors to use the markup
\verb|\proglang{}| (programming languages and large programmable systems),
\verb|\pkg{}| (software packages), \verb|\code{}| (functions, commands,
arguments, etc.). If there is such markup in (sub)section titles (as above), a
plain text version has to be provided in the {\LaTeX} command as well. Below we
also illustrate how abbreviations should be introduced and citation commands can
be employed. See the {\LaTeX} code for more details.
\end{leftbar}

Modeling count variables is a common task in economics and the social sciences.
The classical Poisson regression model for count data is often of limited use in
these disciplines because empirical count data sets typically exhibit
overdispersion and/or an excess number of zeros. The former issue can be
addressed by extending  the plain Poisson regression model in various
directions: e.g., using sandwich covariances or estimating an additional
dispersion parameter (in a so-called quasi-Poisson model). Another more formal
way is to use a negative binomial (NB) regression. All of these models belong to
the family of generalized linear models (GLMs). However, although these models
typically can capture overdispersion rather well, they are in many applications
not sufficient for  modeling excess zeros. Since \citep{Mullahy:1986} there is
increased interest in zero-augmented models that address this issue by a second
model component capturing zero counts. An overview of count data models in
econometrics, including  hurdle and zero-inflated models, is provided in
\citep{Cameron+Trivedi:2013}.

In \proglang{R} \citep{R}, GLMs are provided by the model fitting functions
\fct{glm} in the \pkg{stats} package and \fct{glm.nb} in the \pkg{MASS} package
\citep[][Chapter~7.4]{Venables+Ripley:2002} along with associated methods for
diagnostics and inference. The manuscript that this document is based on
\citep{Zeileis+Kleiber+Jackman:2008} then introduced hurdle and zero-inflated
count models in the functions \fct{hurdle} and \fct{zeroinfl} in the \pkg{pscl}
package \citep{Jackman:2015}. Of course, much more software could be discussed
here, including (but not limited to) generalized additive models for count data
as available in the \proglang{R} packages \pkg{mgcv} \citep{Wood:2006}, 
\pkg{gamlss} \citep{Stasinopoulos+Rigby:2007}, or \pkg{VGAM} \citep{Yee:2009}.
\fi

%% - Manuscript ---------------------

%% - In principle "as usual" again.
%% - When using equations (e.g., {equation}, {eqnarray}, {align}, etc.
%%   avoid empty lines before and after the equation (which would signal a new
%%   paragraph.
%% - When describing longer chunks of code that are _not_ meant for execution
%%   (e.g., a function synopsis or list of arguments), the environment {Code}
%%   is recommended. Alternatively, a plain {verbatim} can also be used.
%%   (For executed code see the next section.)

\section{Methodology and software} \label{sec:methods}


\begin{figure*}[!htb]
%\begin{mdframed}
%  \resizebox{\textwidth}{!}{
         \begin{center}
%
        \subcaptionbox{Activating pathway structure  \label{fig:simple_graph:first}}{%
            \fbox{
            \includegraphics[width=0.45\textwidth]{{simple_graph.png}}
            }
        }%
        \subcaptionbox{Pathway structure with inhibitions \label{fig:simple_graph:second}}{%
            \fbox{
           \includegraphics[width=0.45\textwidth]{{simple_graph_inhibiting.png}}
           }
        }%
%
    \end{center}
   \caption[Simulated graph structures]{\small \textbf{\textbf{Simulated graph structures.}} A constructed graph structure used as an example to demonstrate the simulation procedure in Figures~\ref{fig:simulation_activating} and~\ref{fig:simulation_inhibiting}. Activating links are denoted by black arrows and inhibiting links by red edges. Inhibiting edges have been highlighted in red.}
%}
\label{fig:simple_graph}
%\end{mdframed}
\end{figure*}



Here we present a procedure to simulate gene expression data with correlation structure derived from a known graph structure.  This procedure assumes that transcriptomic data have been generated and follow a log-normal distribution (i.e., $log(X_{ij}) \sim MVN({\bf\mu}, \Sigma)$, where ${\bf\mu}$ and $\Sigma$ are the mean vector  and variance-covariance matrix respectively, for gene expression data derived from a biological pathway) after appropriate normalisation \citep{Law2014, Li2015}. Log-normality of gene expression matches the assumptions of the popular \pkg{limma} package, which is often used for the analysis of intensity-based data from gene expression microarray studies and count-based data from RNA-Seq experiments. This approach has also been applied for modelling UMI-based count data from single-cell RNA-Seq experiments in the \pkg{DESCEND} package \citep{Wang2018}.
%rather than correlated blocks for $\Sigma$ in \pkg{mvtnorm} \citet{mvtnorm}
%This procedure produces a ``Sigma'' variance-covariance matrix ($\Sigma$) for sampling from a multivariate normal distribution (using the \pkg{mvtnorm} \proglang{R} package).

In order to simulate transcriptomic data, a pathway is first constructed as a graph structure, using the \pkg{igraph} \proglang{R} package \citep{igraph}, with the status of the edge relationships defined (i.e, whether they activate or inhibit downstream pathway members). 
%
\textcolor{black}{This procedure uses} a graph structure such as that presented in Figure~\ref{fig:simple_graph:first}. The graph can be defined by an adjacency matrix, \textbf{$A$} (with elements \textcolor{black}{$A_{ij}$}), where  \textcolor{black}{
%
\[
A_{ij} = 
\begin{dcases*}
   1                         & if genes $i$ and $j$ are adjacent \\
   0                         & otherwise
\end{dcases*}
\]}
%
\noindent
A matrix, \textbf{$R$}, with elements \textcolor{black}{$R_{ij}$}, is calculated based on distance (i.e., number of edges contained in the shortest path) between nodes, such that closer nodes are given more weight than more distant nodes, to define inter-node relationships. A geometrically-decreasing (relative) distance weighting is used to achieve this: \textcolor{black}{
%
\[
R_{ij} = 
\begin{dcases*}
   1                            & if genes $i$ and $j$ are adjacent \\
   (\sfrac{1}{2})^{d_{ij}}      & if a path can be found  between genes $i$ and $j$ \\
   0                            & if no path exists between genes $i$ and $j$ 
\end{dcases*}
\]}
%
\noindent
where $d_{ij}$ is the length of the shortest path (i.e., minimum number of edges traversed) between genes (nodes) $i$ and $j$ in graph $G$. Each more distant node is thus related by $\sfrac{1}{2}$ compared to the next nearest, as shown in Figure~\ref{fig:simulation_activating:second}.
An arithmetically-decreasing (absolute) distance weighting is also supported in the \pkg{graphsim} \proglang{R} package which implements this procedure: \textcolor{black}{
%
\[
R_{ij} = 
\begin{dcases*}
   1                            & if genes $i$ and $j$ are adjacent \\
   1-\frac{d_{ij}}{diam(G)}     & if a path can be found  between genes $i$ and $j$ \\
   0                            & if no path exists between genes $i$ and $j$ 
\end{dcases*}
\]
%
}


Assuming a unit variance for each gene, these values can be used to derive a $\Sigma$ matrix:
%
\[
\Sigma_{ij} = 
\begin{dcases*}
   1                            & if $i=j$ \\
   \rho R_{ij}  & otherwise
\end{dcases*}
\]
%
\noindent
where $\rho$ is the correlation between adjacent nodes. 
%The $\Sigma$ matrix (with a diagonal of $1$) would thus have a standard deviation of 1 for each node. 
Thus covariances between adjacent nodes are assigned by a correlation parameter ($\rho$) and the remaining off-diagonal values in the matrix are based on scaling these correlations by the geometrically weighted relationship matrix (or the nearest positive definite matrix for $\Sigma$ with negative correlations).\\

Computing the nearest positive definite matrix is necessary to ensure that the variance-covariance matrix could be inverted when used as a parameter in multivariate normal simulations, particularly when negative correlations are included for inhibitions (as shown below). Matrices that could not be inverted occurred rarely with biologically plausible graph structures but this approach allows for the computation of a plausible correlation matrix when the graph structure given is incomplete or contains loops. When required, the nearest positive definite matrix is computed using the \texttt{nearPD} function of the \pkg{Matrix} \proglang{R} package \citep{Matrix} to perform Higham's algorithm \citep{Higham2002} on variance-covariance matrices. The \pkg{graphsim} package gives a warning when this occurs.

\section{Illustrations} \label{sec:illustrations}
\subsection{Generating a Graph Structure} \label{sec:plot_graph}

The graph structure in Figure~\ref{fig:simple_graph:first} was used to simulate correlated gene expression data by sampling from a multivariate normal distribution using the \pkg{mvtnorm} \proglang{R} package \citep{Genz2009, mvtnorm}.
%This simulation procedure will be demonstrated with the relatively simple constructed graph structure shown in Figure~\ref{fig:simple_graph}. 
The graph structure visualisation in Figure~\ref{fig:simple_graph} was specifically developed for (directed) iGraph objects in \proglang{R} and is available in the \pkg{plot.igraph} and \pkg{igraph.extensions} packages. The \texttt{plot\_directed} function enables customisation of plot parameters for each node or edge, and mixed (directed) edge types for indicating activation or inhibition. These inhibition links (which occur frequently in biological pathways) are demonstrated in Figure~\ref{fig:simple_graph:second}.

A graph structure can be generated and plotted using the following commands in R:


\begin{CodeChunk}
\begin{CodeInput}
#install packages required (once per machine)
install.packages("igraph")
install.packages("devtools")
library("devtools")
devtools::install\_github("TomKellyGenetics/graphsim")
#install custom plotting package
devtools::install\_github("TomKellyGenetics/plot.igraph")
\end{CodeInput}
\end{CodeChunk}

\begin{CodeChunk}
\begin{CodeInput}
#load required packages (once per R instance)
library("igraph")
library("graphsim")
library("plot.igraph")
\end{CodeInput}
\end{CodeChunk}

\begin{CodeChunk}
\begin{CodeInput}
#generate graph structure
graph\_edges <- rbind(c("A", "C"), c("B", "C"), c("C", "D"), c("D", "E"), 
    c("D", "F"), c("F", "G"), c("F", "I"), c("H", "I"))
graph <- graph.edgelist(graph\_edges, directed = T)

#plot graph structure (Figure 1)
plot\_directed(graph, state = "activating", layout = layout.kamada.kawai, 
    cex.node=3, cex.arrow=5, arrow\_clip = 0.2)
\end{CodeInput}
\end{CodeChunk}

\begin{CodeChunk}
\begin{CodeInput}
#generate parameters for inhibitions
state <-  c(1, 1, -1, 1, 1, 1, 1, -1, 1)

#plot graph structure with inhibitions (Figure 2)
plot\_directed(graph, state=state,  layout = layout.kamada.kawai, 
    cex.node=3, cex.arrow=5, arrow\_clip = 0.2)
\end{CodeInput}
\end{CodeChunk}



\subsection{Generating a Simulated Expression Dataset} \label{sec:graphsim_demo}

\begin{figure*}[!hp]
%\begin{mdframed}
%  \resizebox{\textwidth}{!}{
         \begin{center}
%
        \subcaptionbox{Activating pathway structure \label{fig:simulation_activating:first}}{%
            \includegraphics[width=0.35\textwidth]{{simple_graph.png}}
        }%
        \subcaptionbox{Relationship matrix \label{fig:simulation_activating:second}}{%
            \includegraphics[width=0.35\textwidth]{{dist_mat.png}}
        }%
        
        \subcaptionbox{$\Sigma$ (covariance matrix) \label{fig:simulation_activating:third}}{%
           \includegraphics[width=0.35\textwidth]{{sigma_mat.png}}
        }%
	\subcaptionbox{Simulated correlation\label{fig:simulation_activating:fifth}}{%
           \includegraphics[width=0.35\textwidth]{{expr_cor_mat.png}}
        }%
        	
	\subcaptionbox{Simulated expression data (log scale) \label{fig:simulation_activating:fourth}}{%
            \includegraphics[width=1\textwidth]{{expr_mat_wide.png}}
        }%
        %\subcaptionbox{Simulated gene function calls \label{fig:simulation_activating:sixth}}{%
        %   \includegraphics[width=0.35\textwidth]{{expr_disc_mat.png}}
        %}%
    \end{center}
   \caption[Simulating expression from a graph structure]{\small \textbf{\textbf{Simulating expression from a graph structure.}} An example of a graph structure (a) that has been used to derive a relationship matrix (b), $\Sigma$ matrix (c) and correlation structure (d) from the relative distances between the nodes. Non-negative values are coloured white to red from $0$ to $1$. This $\Sigma$ matrix has been used to generate a simulated expression dataset of 100 samples (coloured blue to red from low to high) via sampling from the multivariate normal distribution. Here genes with closer relationships in the pathway structure show higher correlation between simulated values.}
%}
\label{fig:simulation_activating}
%\end{mdframed}
\end{figure*}



The correlation parameter of $\rho = 0.8$ is used to demonstrate the inter-correlated datasets using a geometrically-generated relationship matrix (as used for the example in Figure~\ref{fig:simulation_activating:third}). This $\Sigma$ matrix was then used to sample from a multivariate normal distribution such that each gene had a mean of $0$, standard deviation $1$, and covariance within the range $[0,1]$ so that the off-diagonal elements of $\Sigma$ represent correlations. This procedure generated a simulated (continuous normally-distributed) log-expression profile for each node (Figure~\ref{fig:simulation_activating:fourth}) with a corresponding correlation structure (Figure~\ref{fig:simulation_activating:fifth}). The simulated correlation structure closely resembled the expected correlation structure ($\Sigma$ in Figure~\ref{fig:simulation_activating:third}) even for the relatively modest sample size ($N=100$) illustrated in Figure~\ref{fig:simulation_activating}. Once a gene expression dataset comprising multiple pathways has been generated (as in Figure~\ref{fig:simulation_activating:fourth}), 
%then a discrete matrix of gene function was constructed with a functional threshold quantile to simulate functional relationships of synthetic lethality (Figure~\ref{fig:SL_Model_Expression}). %Throughout this thesis, this threshold is the 0.3 quantile (as discussed in Section~\ref{methods:SL_Model}) which generates functional discrete matrices such as those used for synthetic lethal simulation in Section~\ref{methods:simulating_SL} (as shown Figure~\ref{fig:simulation_activating:sixth}).
it can then be used to test procedures designed for analysis of empirical gene expression data (such as those generated by microarrays or RNA-Seq) that have been normalised on a log-scale.

The simulated dataset can be generated using the following \proglang{R} code:

\begin{CodeChunk}
\begin{CodeInput}
#adjacency matrix
adj\_mat <- make\_adjmatrix\_graph(graph)
  
#relationship matrix
dist\_mat <- make\_distance\_graph(graph\_test4, absolute = F)

#sigma matrix directly from graph
sigma\_mat <- make\_sigma\_mat\_dist\_graph(graph, 0.8, absolute = F)

#show shortest paths of graph
shortest\_paths <- shortest.paths(graph)

#generate expression data directly from graph
expr <- generate\_expression(100, graph, cor = 0.8, mean = 0, comm = F,
                            dist = T, absolute = F, state = state)
\end{CodeInput}
\end{CodeChunk}

\begin{CodeChunk}
\begin{CodeInput}
##plot steps

#plot adjacency matrix
heatmap.2(make\_adjmatrix\_graph(graph), scale = "none", trace = "none", 
          col = colorpanel(3, "grey75", "white", "blue"),
          colsep = 1:length(V(graph)), rowsep = 1:length(V(graph)))

#plot relationship matrix
heatmap.2(make\_distance\_graph(graph\_test4, absolute = F), scale = "none",
          trace = "none", col = bluered(50), colsep = 1:length(V(graph)), 
        rowsep = 1:length(V(graph)))

#plot sigma matrix
heatmap.2(make\_sigma\_mat\_dist\_graph(graph, 0.8, absolute = F), scale = "none", 
          trace = "none", col = bluered(50), colsep = 1:length(V(graph)), 
          rowsep = 1:length(V(graph)))
expr <- generate\_expression(100, graph, cor = 0.8, mean = 0, comm = F, dist = T,
                            absolute = F, state = state)

#plot simulated expression data
heatmap.2(expr, scale = "none", trace = "none", col = bluered(50), 
          colsep = 1:length(V(graph)), rowsep = 1:length(V(graph)))

#plot simulated correlations
heatmap.2(cor(t(expr)), scale = "none", trace = "none", col = bluered(50), 
          colsep = 1:length(V(graph)), rowsep = 1:length(V(graph)))
\end{CodeInput}
\end{CodeChunk}

\begin{figure*}[!hp]
%\begin{mdframed}
%  \resizebox{\textwidth}{!}{
         \begin{center}
%
        \subcaptionbox{Inhibiting pathway structure\label{fig:simulation_inhibiting:first}}{%
            \includegraphics[width=0.35\textwidth]{{simple_graph_inhibiting.png}}
        }%
        \subcaptionbox{Relationship matrix \label{fig:simulation_inhibiting:second}}{%
            \includegraphics[width=0.35\textwidth]{{dist_mat.png}}
        }%
        
        \subcaptionbox{$\Sigma$ (covariance matrix) \label{fig:simulation_inhibiting:third}}{%
           \includegraphics[width=0.35\textwidth]{{sigma_mat_inhibiting.png}}
        }%
	\subcaptionbox{Simulated correlation\label{fig:simulation_inhibiting:fifth}}{%
           \includegraphics[width=0.35\textwidth]{{expr_cor_mat_inhibiting.png}}
        }%
        
        \subcaptionbox{Simulated expression data (log scale) \label{fig:simulation_inhibiting:fourth}}{%
            
            \includegraphics[width=1\textwidth]{{expr_mat_inhibiting_wide.png}}
        }%
        %\subcaptionbox{Simulated gene function calls \label{fig:simulation_inhibiting:sixth}}{%
        %   \includegraphics[width=0.35\textwidth]{{expr_inhib_disc_mat.png}}
        %}%
    \end{center}
    \caption[Simulating expression from graph structure with inhibitions]{\small \textbf{\textbf{Simulating expression from graph structure with inhibitions.}} An example of a graph structure (a), that has been used to derive a relationship matrix (b), $\Sigma$ matrix (c), and correlation structure (d), from the relative distances between the nodes. These values are coloured blue to red from $-1$ to $1$. This has been used to generate a simulated expression dataset of 100 samples (coloured blue to red from low to high) via sampling from the multivariate normal distribution. Here the inhibitory relationships between genes are reflected in negatively correlated simulated  values.}
%}
\label{fig:simulation_inhibiting}
%\end{mdframed}
\end{figure*}


The simulation procedure (Figure~\ref{fig:simulation_activating}) can similarly be used for pathways containing inhibitory links (Figure~\ref{fig:simulation_inhibiting}) with several refinements. With the inhibitory links (Figure~\ref{fig:simulation_inhibiting:first}), distances are calculated in the same manner as before (Figure~\ref{fig:simulation_inhibiting:second}) with inhibitions accounted for by iteratively multiplying downstream nodes by $-1$ to form modules with negative correlations between them (Figures~\ref{fig:simulation_inhibiting:third} and~\ref{fig:simulation_inhibiting:fifth}). A multivariate normal distribution with these negative correlations can be sampled to generate simulated data (Figure~\ref{fig:simulation_inhibiting:fourth}).% and~\ref{fig:simulation_inhibiting:sixth}).  


\begin{figure*}[!hp]
%\begin{mdframed}
%  \resizebox{\textwidth}{!}{
         \begin{center}
%
        \subcaptionbox{TGF-$\beta$ activates SMADs\label{fig:simulation_smad:first}}{%
            \includegraphics[width=0.35\textwidth]{{graph_smad_simple.pdf}}
        }%
        \subcaptionbox{Relationship matrix \label{fig:simulation_smad:second}}{%
            \includegraphics[width=0.35\textwidth]{{dist_mat_smad_simple.png}}
        }%
        
        \subcaptionbox{$\Sigma$ (covariance matrix) \label{fig:simulation_smad:third}}{%
           \includegraphics[width=0.35\textwidth]{{sigma_mat_inhibiting_smad_simple.png}}
        }%
	\subcaptionbox{Simulated correlation\label{fig:simulation_smad:fifth}}{%
           \includegraphics[width=0.35\textwidth]{{expr_cor_mat_inhibiting_smad_simple.png}}
        }%
        
        \subcaptionbox{Simulated expression data (log scale) \label{fig:simulation_smad:fourth}}{%
            
            \includegraphics[width=0.8\textwidth]{{expr_mat_inhibiting_wide_smad_simple.png}}
        }%
        %\subcaptionbox{Simulated gene function calls \label{fig:simulation_smad:sixth}}{%
        %   \includegraphics[width=0.35\textwidth]{{expr_inhib_disc_mat.png}}
        %}%
    \end{center}
   \caption[Simulating expression from a biological pathway graph structure]{\small \textbf{\textbf{Simulating expression from graph structure with inhibitions.}} The graph structure (a) of a known biological pathway, the TGF-$\beta$ receptor signaling activates SMADs (R-HSA-2173789), was used to derive a relationship matrix (b), $\Sigma$ matrix (c) and correlation structure (d) from the relative distances between the nodes. These values are coloured blue to red from $-1$ to $1$. This has been used to generate a simulated expression dataset of 100 samples (coloured blue to red from low to high) via sampling from the multivariate normal distribution. Here modules of genes with correlated expression can be clearly discerned.}
%}
\label{fig:simulation_smad}
%\end{mdframed}
\end{figure*}

The simulation procedure is also demonstrated here (Figure~\ref{fig:simulation_smad}) on a pathway structure for a known biological pathway (from reactome R-HSA-2173789) of TGF-$\beta$ receptor signaling activates SMADs (Figure~\ref{fig:simulation_smad:first}) derived from the Reactome database version 52 \citep{Reactome}. Distances are calculated in the same manner as before (Figure~\ref{fig:simulation_smad:second}) producing blocks of correlated genes (Figures~\ref{fig:simulation_inhibiting:third} and~\ref{fig:simulation_inhibiting:fifth}). This shows that multivariate normal distribution can be sampled to generate simulated data to represent expression with the complexity of a biological pathway (Figure~\ref{fig:simulation_inhibiting:fourth}). Here \textit{SMAD7} exhibits negative correlations with the other SMADs consistent with it's functions as as an ``inhibitor SMAD'' with competitively inhibits \textit{SMAD4}. % and~\ref{fig:simulation_inhibiting:sixth}).  

\iffalse
These simulated datasets could then be used for simulating synthetic lethal partners of a query gene within a graph network. The query gene was assumed to be separate from the graph network pathway and was added to the dataset using the procedure in Section~\ref{methods:simulating_SL}. Thus I can simulate known synthetic lethal partner genes within a synthetic lethal partner pathway structure.
\fi

\iffalse

	\includegraphics{{dist_mat.png}}
	\includegraphics{{sigma_mat.png}}
		\includegraphics{{expr_mat.png}}
			\includegraphics{{expr_cor_mat.png}}
		\includegraphics{{expr_disc_mat.png}}
		
	\includegraphics{{state_matrix_inhibiting.png}}
	\includegraphics{{dist_mat.png}}
		\includegraphics{{sigma_mat_inhibiting.png}}
		\includegraphics{{expr_inhib_mat.png}}
			\includegraphics{{expr_inhib_cor_mat.png}}
		\includegraphics{{expr_inhib_disc_mat.png}}
	
\fi	

%\FloatBarrier


%% - Illustrations --------------------

%% - Virtually all JSS manuscripts list source code along with the generated
%%   output. The style files provide dedicated environments for this.
%% - In R, the environments {Sinput} and {Soutput} - as produced by Sweave() or
%%   or knitr using the render_sweave() hook - are used (without the need to
%%   load Sweave.sty).
%% - Equivalently, {CodeInput} and {CodeOutput} can be used.
%% - The code input should use "the usual" command prompt in the respective
%%   software system.
%% - For \proglang{R} code, the prompt "R> " should be used with "+  " as the
%%   continuation prompt.
%% - Comments within the code chunks should be avoided - these should be made
%%   within the regular LaTeX text.

\iffalse

\section{Illustrations} \label{sec:illustrations}

For a simple illustration of basic Poisson and NB count regression the
\code{quine} data from the \pkg{MASS} package is used. This provides the number
of \code{Days} that children were absent from school in Australia in a
particular year, along with several covariates that can be employed as regressors.
The data can be loaded by
%
\begin{CodeChunk}
\begin{CodeInput}
R> data("quine", package = "MASS")
\end{CodeInput}
\end{CodeChunk}
%
and a basic frequency distribution of the response variable is displayed in
Figure~\ref{fig:quine}.

\begin{leftbar}
For code input and output, the style files provide dedicated environments.
Either the ``agnostic'' \verb|{CodeInput}| and \verb|{CodeOutput}| can be used
or, equivalently, the environments \verb|{Sinput}| and \verb|{Soutput}| as
produced by \fct{Sweave} or \pkg{knitr} when using the \code{render_sweave()}
hook. Please make sure that all code is properly spaced, e.g., using
\code{y = a + b * x} and \emph{not} \code{y=a+b*x}. Moreover, code input should
use ``the usual'' command prompt in the respective software system. For
\proglang{R} code, the prompt \code{"R> "} should be used with \code{"+  "} as
the continuation prompt. Generally, comments within the code chunks should be
avoided - and made in the regular {\LaTeX} text instead. Finally, empty lines
before and after code input/output should be avoided (see above).
\end{leftbar}

\begin{figure}[t!]
\centering
\includegraphics{article-visualization}
\caption{\label{fig:quine} Frequency distribution for number of days absent
from school.}
\end{figure}

As a first model for the \code{quine} data, we fit the basic Poisson regression
model. (Note that JSS prefers when the second line of code is indented by two
spaces.)
%
\begin{CodeChunk}
\begin{CodeInput}
R> m\_pois <- glm(Days ~ (Eth + Sex + Age + Lrn)^2, data = quine,
+    family = poisson)
\end{CodeInput}
\end{CodeChunk}
%
To account for potential overdispersion we also consider a negative binomial
GLM.
%
\begin{CodeChunk}
\begin{CodeInput}
R> library("MASS")
R> m\_nbin <- glm.nb(Days ~ (Eth + Sex + Age + Lrn)^2, data = quine)
\end{CodeInput}
\end{CodeChunk}
%
In a comparison with the BIC the latter model is clearly preferred.
%
\begin{CodeChunk}
\begin{CodeInput}
R> BIC(m\_pois, m\_nbin)
\end{CodeInput}
\begin{CodeOutput}
       df      BIC
m\_pois 18 2046.851
m\_nbin 19 1157.235
\end{CodeOutput}
\end{CodeChunk}
%
Hence, the full summary of that model is shown below.
%
\begin{CodeChunk}
\begin{CodeInput}
R> summary(m\_nbin)
\end{CodeInput}
\begin{CodeOutput}
Call:
glm.nb(formula = Days ~ (Eth + Sex + Age + Lrn)^2, data = quine, 
    init.theta = 1.60364105, link = log)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-3.0857  -0.8306  -0.2620   0.4282   2.0898  

Coefficients: (1 not defined because of singularities)
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  3.00155    0.33709   8.904  < 2e-16 ***
EthN        -0.24591    0.39135  -0.628  0.52977    
SexM        -0.77181    0.38021  -2.030  0.04236 *  
AgeF1       -0.02546    0.41615  -0.061  0.95121    
AgeF2       -0.54884    0.54393  -1.009  0.31296    
AgeF3       -0.25735    0.40558  -0.635  0.52574    
LrnSL        0.38919    0.48421   0.804  0.42153    
EthN:SexM    0.36240    0.29430   1.231  0.21818    
EthN:AgeF1  -0.70000    0.43646  -1.604  0.10876    
EthN:AgeF2  -1.23283    0.42962  -2.870  0.00411 ** 
EthN:AgeF3   0.04721    0.44883   0.105  0.91622    
EthN:LrnSL   0.06847    0.34040   0.201  0.84059    
SexM:AgeF1   0.02257    0.47360   0.048  0.96198    
SexM:AgeF2   1.55330    0.51325   3.026  0.00247 ** 
SexM:AgeF3   1.25227    0.45539   2.750  0.00596 ** 
SexM:LrnSL   0.07187    0.40805   0.176  0.86019    
AgeF1:LrnSL -0.43101    0.47948  -0.899  0.36870    
AgeF2:LrnSL  0.52074    0.48567   1.072  0.28363    
AgeF3:LrnSL       NA         NA      NA       NA    
-
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for Negative Binomial(1.6036) family taken to be 1)

    Null deviance: 235.23  on 145  degrees of freedom
Residual deviance: 167.53  on 128  degrees of freedom
AIC: 1100.5

Number of Fisher Scoring iterations: 1


              Theta:  1.604 
          Std. Err.:  0.214 

 2 x log-likelihood:  -1062.546 
\end{CodeOutput}
\end{CodeChunk}
\fi


%% - Summary/conclusions/discussion ---------------

\FloatBarrier

\section{Summary and discussion} \label{sec:summary}

\begin{leftbar}
Biological pathways are of fundamental importance to understanding molecular biology. In order to translate findings from genomics studies into real-world applications such as improved healthcare, the roles of genes must be studied in the context of molecular pathways. Here we present a statistical framework to simulate gene expression from biological pathways, and provide the \pkg{graphsim} package in \proglang{R} to generate these simulated datasets. This approach is versatile and can be fine-tuned for modelling existing biological pathways or for testing whether constructed pathways can be detected by other means. In particular, methods to infer biological pathways and gene regulatory networks from gene expression data can be tested on simulated datasets using this framework. The package also enables simulation of complex gene expression datasets to test how these pathways impact on statistical analysis of gene expression data using existing methods or novel statistical methods being developed for gene expression data analysis. 
\end{leftbar}


%% - Optional special unnumbered sections -------------

\section*{Computational details}

\iffalse
\begin{leftbar}
If necessary or useful, information about certain computational details
such as version numbers, operating systems, or compilers could be included
in an unnumbered section. Also, auxiliary packages (say, for visualizations,
maps, tables, \dots) that are not cited in the main text can be credited here.
\end{leftbar}
\fi

The results in this paper were obtained using
\proglang{R}~3.6.1 with the
\pkg{igraph}~1.2.4.1 \pkg{Matrix}~1.2-17, \pkg{matrixcalc}~1.0-3, and \pkg{mvtnorm}~1.0-11 packages. \proglang{R} itself
and all dependent packages used are available from the Comprehensive
\proglang{R} Archive Network (CRAN) at
\url{https://CRAN.R-project.org/}. 
The \pkg{graphsim} and \pkg{plot.igraph} packages presented can be installed from \url{https://github.com/TomKellyGenetics/graphsim} 
and \url{https://github.com/TomKellyGenetics/plot.igraph} respectively. These functions can also be installed using the \pkg{igraph.extensions} library at \url{https://github.com/TomKellyGenetics/igraph.extensions} which includes other plotting functions used. This software is cross-platform and compatible with \proglang{R} installations on Windows, Mac, and Linux operating systems. The package GitHub repository also contains Vignettes with more information and examples on running functions released in the \proglang{R} package. The package (\pkg{graphsim}~0.1.0) meets CRAN submission criteria and will be released.


\section*{Acknowledgements} %change to British spelling from American: ``Acknowledgments''

\begin{leftbar}
This package was developed as part of a PhD research project funded by the Postgraduate Tassell Scholarship in Cancer Research Scholarship awarded to STK. We thank members of the Laboratory of Professor Satoru Miyano at the University of Tokyo, Institute for Medical Science, Professor Seiya Imoto, Associate Professor Rui Yamaguchi, and Dr Paul Sheridan (Assistant Professor at Hirosaki University,CSO at Tupac Bio) for helpful discussions in this field. We also thank Professor Parry Guilford at the University of Otago, Professor Cristin Print at the University of Auckland, and Dr Erik Arner at the RIKEN Center for Integrative Medical Sciences for their excellent advice during this project.
\end{leftbar}


%% - Bibliography ---------------------
%% - References need to be provided in a .bib BibTeX database.
%% - All references should be made with \cite, \citet, \cite, \citealp etc.
%%   (and never hard-coded). See the FAQ for details.
%% - JSS-specific markup (\proglang, \pkg, \code) should be used in the .bib.
%% - Titles in the .bib should be in title case.
%% - DOIs should be included where available.

\bibliography{refs}


%% - Appendix (if any) -------------------
%% - After the bibliography with page break.
%% - With proper section titles and _not_ just "Appendix".

\newpage

%\begin{appendix}

%\section{More technical details} \label{app:technical}

\begin{leftbar}
%Would it be appropriate to include the Vignette for the package here?
\iffalse
Appendices can be included after the bibliography (with a page break). Each
section within the appendix should have a proper section title (rather than
just \emph{Appendix}).

For more technical style details, please check out JSS's style FAQ at
\url{https://www.jstatsoft.org/pages/view/style#frequently-asked-questions}
which includes the following topics:
\begin{itemize}
  \item Title vs.\ sentence case.
  \item Graphics formatting.
  \item Naming conventions.
  \item Turning JSS manuscripts into \proglang{R} package vignettes.
  \item Trouble shooting.
  \item Many other potentially helpful details\dots
\end{itemize}
\fi
\end{leftbar}

\iffalse
\section[Using BibTeX]{Using \textsc{Bib}{\TeX}} \label{app:bibtex}

\begin{leftbar}
References need to be provided in a \textsc{Bib}{\TeX} file (\code{.bib}). All
references should be made with \verb|\cite|, \verb|\citet|, \verb|\citep|,
\verb|\citealp| etc.\ (and never hard-coded). This commands yield different
formats of author-year citations and allow to include additional details (e.g.,
pages, chapters, \dots) in brackets. In case you are not familiar with these
commands see the JSS style FAQ for details.

Cleaning up \textsc{Bib}{\TeX} files is a somewhat tedious task - especially
when acquiring the entries automatically from mixed online sources. However,
it is important that informations are complete and presented in a consistent
style to avoid confusions. JSS requires the following format.
\begin{itemize}
  \item JSS-specific markup (\verb|\proglang|, \verb|\pkg|, \verb|\code|) should
    be used in the references.
  \item Titles should be in title case.
  \item Journal titles should not be abbreviated and in title case.
  \item DOIs should be included where available.
  \item Software should be properly cited as well. For \proglang{R} packages
    \code{citation("pkgname")} typically provides a good starting point.
\end{itemize}
\end{leftbar}
\fi

%\end{appendix}

%% --------------------------


\end{document}
